xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,203 @@
+    List<String> unknownPolicyNames = new ArrayList<>();
+    // Set up activation used in CEL library's eval function.
+    Activation activation = Activation.copyOf(extractFields(args));
+    // Go through each RBAC in the Envoy RBAC list.
+    for (RbacEngine rbacEngine : rbacEngines) {
+      // Go through each condition in the RBAC policy.
+      for (Map.Entry<String, Expr> condition : rbacEngine.conditions.entrySet()) {
+        try {
+          if (matches(condition.getValue(), activation)) {
+            authorizationDecision = rbacEngine.action == Action.ALLOW 
+                ? AuthorizationDecision.Decision.ALLOW : AuthorizationDecision.Decision.DENY;
+            matchingPolicyNames.add(condition.getKey());
+          }
+        } catch (InterpreterException e) {
+          unknownPolicyNames.add(condition.getKey());
+        }
+      }
+      if (matchingPolicyNames.size() > 0) {
+        return new AuthorizationDecision(authorizationDecision, matchingPolicyNames);
+      }
+      if (unknownPolicyNames.size() > 0) {
+        return new AuthorizationDecision(
+            AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+      }
+    }
+    // No RBAC conditions matched and didn't find unknown conditions.
+    if (this.rbacEngines.size() == 1 && this.rbacEngines.get(0).action == Action.DENY) {
+      return new AuthorizationDecision(
+          AuthorizationDecision.Decision.ALLOW, new ArrayList<String>());
+    }
+    return new AuthorizationDecision(AuthorizationDecision.Decision.DENY, new ArrayList<String>());
+  }
+
+  /** Evaluate if a condition matches the given Enovy Attributes using CEL library. */
+  protected boolean matches(Expr condition, Activation activation) throws InterpreterException {
+    // Set up interpreter used in CEL library's eval function.
+    List<Descriptor> descriptors = new ArrayList<>();
+    RuntimeTypeProvider messageProvider = DescriptorMessageProvider.dynamicMessages(descriptors);
+    Dispatcher dispatcher = DefaultDispatcher.create();
+    Interpreter interpreter = new DefaultInterpreter(messageProvider, dispatcher);
+    // Parse the generated result object to a boolean variable.
+    try {
+      Object result = interpreter.createInterpretable(condition).eval(activation);
+      if (result instanceof Boolean) {
+        return Boolean.valueOf(result.toString());
+      }
+      // Throw an InterpreterException if there are missing Envoy Attributes.
+      if (result instanceof IncompleteData) {
+        throw new InterpreterException.Builder("Incomplete Envoy Attributes to be evaluated.")
+            .build(); 
+      }
+    } catch (InterpreterException e) {
+      // If any InterpreterExceptions are catched, throw it and log the error.
+      log.log(Level.WARNING, e.toString(), e);
+      throw e;
+    }
+    return false;
+  }
+
+  /** Extract Envoy Attributes from EvaluateArgs. */
+  protected ImmutableMap<String, Object> extractFields(EvaluateArgs<ReqT, RespT> args) {
```

Comment:
As far as I understand, the design of part is still ongoing yet. In this PR, we can assume evaluate function takes the activation and gRPC stack will set it directly(so in this PR we don't have to deal with that `EvaluateArgs<ReqT, RespT> args` anymore).
@jiangtaoli2016 Feel free to correct me if this is not our current plan. Thanks!

Reply:
Sounds good!

Reply:
why `protected` ? Also related to whether `AuthorizationEngine` can be final or needs to be extended. If final this method can be `final` or `private`

Reply:
Currently there is no open-sourced CEL library in Java, so all the APIs used in Authorization Engine is from a mock Cel Library interface that only holds the function signature. Therefore, I had to use Mockito to mock the behavior of helper functions in order to test the logic of `evaluate` function on top of that. With that being said, all the helper functions are set to protected for now, so that they can be exposed for testing, but we will change it once CEL library is open sourced in the future.